lighthouse-core/lib/tracehouse/trace-processor.js

```diff
@@ -478,17 +480,67 @@ class TraceProcessor {
   }
 
   /**
-   * @param {{candidateEventName: string, invalidateEventName: string, events: LH.TraceEvent[], timeOriginEvt: LH.TraceEvent}} options
-   * @return {{lcp: LH.TraceEvent | undefined, invalidated: boolean}}
+   * @param {LH.TraceEvent} evt
+   * @return {evt is LCPEvent}
    */
-  static computeValidLCP(options) {
-    const {
-      candidateEventName,
-      invalidateEventName,
-      events,
-      timeOriginEvt,
-    } = options;
+  static isLCPEvent(evt) {
+    if (evt.name !== 'largestContentfulPaint::Invalidate' &&
+        evt.name !== 'largestContentfulPaint::Candidate') return false;
+    return !!(evt.args && evt.args.frame);
+  }
+
+  /**
+   * @param {LH.TraceEvent} evt
+   * @return {evt is LCPCandidateEvent}
+   */
+  static isLCPCandidateEvent(evt) {
+    if (evt.name !== 'largestContentfulPaint::Candidate') return false;
+    return !!(evt.args && evt.args.frame && evt.args.data && evt.args.data.size !== undefined);
+  }
+
+  /**
+   * @param {LH.TraceEvent[]} events
+   * @param {LH.TraceEvent} timeOriginEvent
+   * @return {LCPEvent | undefined}
+   */
+  static computeValidLCPAllFrames(events, timeOriginEvent) {
```

Comment:
I'm having a bit of trouble understanding how this differs from the main frame function. Is it possible to have the "all frames" utilize the "main frame" function (it'd pick the maximum LCP across ass)? Is that functionally the same?

Reply:
LCP is computed on a per-frame basis, so looking at the final LCP event will only give you the LCP for one frame.

The reason I didn't use `computeValidLCP` in this function was because it would require filtering `events` for every single frame. It seemed better to just do a single pass, tracking the LCP for any frames we find.

Taking a second look, I think the main frame function can be deprecated and we could just return the main frame LCP alongside the LCP from all frames in this function.